packages/lodestar/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -0,0 +1,310 @@
+    const forkName = state.config.getForkName(state.slot);
+    if (forkName === ForkName.phase0) {
+      return this.getAttestationsForPhase0Block(state, isValidAttestation);
+    }
+    return this.getAttestationsForAltairBlock(state, isValidAttestation);
+  }
+
+  /**
+   * Get all attestations optionally filtered by `attestation.data.slot`
+   * @param bySlot slot to filter, `bySlot === attestation.data.slot`
+   */
+  getAll(state: CachedBeaconState<allForks.BeaconState>, bySlot?: Slot): phase0.Attestation[] {
+    // we don't want to validate attestations
+    let attestations = this.getAttestationsForBlock(state, true);
+    if (bySlot) {
+      attestations = attestations.filter((att) => att.data.slot === bySlot);
+    }
+    return attestations;
+  }
+
+  /**
+   * Get attestations to be included in a phase0 block.
+   * As we are close to altair, this is not really important, it's mainly for e2e.
+   * The performance is not great due to the different BeaconState data structure to altair.
+   */
+  private getAttestationsForPhase0Block(
+    state: CachedBeaconState<allForks.BeaconState>,
+    isValidAttestation = false
+  ): phase0.Attestation[] {
+    if (state.config.getForkName(state.slot) !== ForkName.phase0) {
+      throw new Error(`Not support proposing block at slot ${state.slot}`);
+    }
+    const phase0State = state as CachedBeaconState<phase0.BeaconState>;
+    const {epochCtx} = phase0State;
+    const currentEpoch = computeEpochAtSlot(state.slot);
+    const extractParticipation = (attestations: List<phase0.PendingAttestation>): Set<ValidatorIndex> => {
+      const allParticipants = new Set<ValidatorIndex>();
+      for (const att of readonlyValues(attestations)) {
+        const aggregationBits = att.aggregationBits;
+        const attData = att.data;
+        const attSlot = attData.slot;
+        const committeeIndex = attData.index;
+        const committee = epochCtx.getBeaconCommittee(attSlot, committeeIndex);
+        const participants = zipIndexesCommitteeBits(committee, aggregationBits);
+        for (const participant of participants) {
+          allParticipants.add(participant);
+        }
+      }
+      return allParticipants;
+    };
+    const previousEpochParticipants = extractParticipation(phase0State.previousEpochAttestations);
+    const currentEpochParticipants = extractParticipation(phase0State.currentEpochAttestations);
+    const slots = Array.from(this.attestationGroupByDataHashBySlot.keys()).sort((a, b) => b - a);
+    const attestations: phase0.Attestation[] = [];
+    for (const slot of slots) {
+      const attestationGroupByDataHash = this.attestationGroupByDataHashBySlot.get(slot);
+      // should not happen
+      if (!attestationGroupByDataHash) {
+        throw Error(`No aggregated attestation pool for slot=${slot}`);
+      }
+      const attestationGroups = Array.from(attestationGroupByDataHash.values());
+      const epoch = computeEpochAtSlot(slot);
+      const participants =
```

Comment:
It necessary to compute this `participants` variable on every slot of the loop?

Reply:
each slot may belong to either previous epoch or current epoch so we have to compute it here, it's just a `if/else` clause here since `currentEpochParticipants` and `previoiusEpochParticipants` are precomputed once before the loop.

we should focus more on the same function for altair btw